src/core/server/rendering/rendering_service.tsx

```diff
@@ -0,0 +1,126 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React from 'react';
+import { renderToStaticMarkup } from 'react-dom/server';
+import { format } from 'prettier';
+import { take } from 'rxjs/operators';
+
+import { i18n } from '@kbn/i18n';
+
+import { CoreService } from '../../types';
+import { CoreContext } from '../core_context';
+import { ensureRawRequest } from '../http/router';
+import { Template } from './views';
+import { RenderingSetupDeps, RenderingServiceSetup } from './types';
+
+/** @internal */
+export class RenderingService implements CoreService<RenderingServiceSetup> {
+  constructor(private readonly coreContext: CoreContext) {}
+
+  public async setup({
+    http,
+    legacy,
+    legacyPlugins,
+    plugins,
+  }: RenderingSetupDeps): Promise<RenderingServiceSetup> {
+    const uiPlugins = await Promise.all(
+      [...plugins.uiPlugins.public].map(async ([pluginId, plugin]) => {
+        const config = ((await plugins.uiPlugins.browserConfigs
+          .get(pluginId)
+          ?.pipe(take(1))
+          .toPromise()) ?? {}) as Record<string, any>;
+
+        return { id: pluginId, plugin, config };
+      })
+    );
+
+    return {
+      getRenderingProvider: ({ request, uiSettings, injectedVarsOverrides }) => {
+        const basePath = http.basePath.get(request);
+
+        return {
+          render: async (
```

Comment:
I'm still unsure why this is two-step process rather than a single function?

Reply:
We need a few things to be able to render the page:

```ts
const renderingSetup = await this.rendering.setup({
  http: httpSetup,
  legacy: this.legacy,
  legacyPlugins,
  plugins: pluginsSetup,
});
```

In addition, we also need access to the request to get the HTTP base path, and creating an API that is contextual for the request means this is a 2-step process, just like all the other client-based APIs:

```ts
const adminClient = await coreSetup.elasticsearch.adminClient$.pipe(take(1)).toPromise();
const dataClient = await coreSetup.elasticsearch.dataClient$.pipe(take(1)).toPromise();
const savedObjectsClient = coreSetup.savedObjects.getScopedClient(req);
const uiSettingsClient = coreSetup.uiSettings.asScopedToClient(savedObjectsClient);
const renderingProvider = await rendering.getRenderingProvider({
  request: req,
  uiSettings: uiSettingsClient,
});
```

Does that make sense?

Reply:
We need a few things to be able to render the page:

```ts
const renderingSetup = await this.rendering.setup({
  http: httpSetup,
  legacy: this.legacy,
  legacyPlugins,
  plugins: pluginsSetup,
});
```

In addition, we also need access to the request to get the HTTP base path, and creating an API that is contextual for the request means this is a 2-step process, just like all the other client-based APIs:

```ts
const adminClient = await coreSetup.elasticsearch.adminClient$.pipe(take(1)).toPromise();
const dataClient = await coreSetup.elasticsearch.dataClient$.pipe(take(1)).toPromise();
const savedObjectsClient = coreSetup.savedObjects.getScopedClient(req);
const uiSettingsClient = coreSetup.uiSettings.asScopedToClient(savedObjectsClient);
const renderingProvider = await rendering.getRenderingProvider({
  request: req,
  uiSettings: uiSettingsClient,
});
```

Does that make sense?